src/v/pandaproxy/schema_registry/json.cc

```diff
@@ -568,26 +591,108 @@ ss::future<document_context_jsoncons> parse_jsoncons(iobuf buf) {
           ? validate_json_schema(maybe_dialect.value(), schema).value()
           : try_validate_json_schema(schema).value();
 
-    // this function will resolve al local ref against their respective baseuri.
-    auto bundled_schemas_map
-      = collect_bundled_schema_and_fix_refs(schema, dialect).value();
-
+    // this function will resolve al local ref against their respective
+    // baseuri. if we are currently in the process of parsing a external
+    // schema, default_id will have a value and will be used as a baseuri if
+    // $id is not set. this is done to ensure that relative refs inside the
+    // external schema do not clash with relative refs in the root schema,
+    // if it does not have a base_uri.
+    auto bundled_schemas_map = collect_bundled_schema_and_fix_refs(
+                                 schema, dialect, default_id)
+                                 .value();
     auto schemas_index = document_context_jsoncons::schemas_index_t{
       std::move_iterator(bundled_schemas_map.begin()),
       std::move_iterator(bundled_schemas_map.end())};
 
+    // now the schema is a valid schema, the $ref are absolute, and the bundled
+    // schemas are collected in schemas_index. proceed to parse the external
+    // refs
+
+    // collect the iobufs for each external schema. all recursive external refs
+    // are collected, there will be no duplicated names
+    auto external_refs = (co_await collect_schema(store, {}, refs)).get();
+
+    auto external_schemas = document_context_jsoncons::external_schemas_map_t{};
+    for (auto& [ref_name, def] : external_refs) {
+        // recursive call to parse the external schema. it will have no external
+        // refs but it will have a default_id and all the relative refs will be
+        // resolved against it
+        auto ref_document = co_await parse_jsoncons(
+          store, std::move(def), ref_name, {});
```

Comment:
Am I reading it correctly that this means that we're not supporting transitive external references (ie. referencing a schema that also references schemas)?